### PR TITLE
Adds books to API

### PR DESCRIPTION
**Pull request by dkocich**
Requires to merge changes in v6_common https://github.com/c2corg/v6_common/pull/33

### CODE REVIEW

File: c2corg_api/scripts/migration/documents/books.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            ' where '
+            ' b.redirects_to is null; '
+        )
+
+    def get_query(self):
+        return (
+            ' select '
+            '   ba.id, ba.document_archive_id, ba.is_latest_version, '
+            '   ba.is_protected, ba.redirects_to, '
+            '   ST_Force2D(ST_SetSRID(ba.geom, 3857)) geom, '
```
Comment (tsauerwein) [inline c2corg_api/scripts/migration/documents/books.py]: Do books have a geometry?

File: c2corg_api/models/book.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    url = Column(String(255))
+    isbn = Column(String(17))
+    book_types = Column(ArrayOfEnum(enums.book_type))
+    nb_pages = Column(SmallInteger)
+
+    # it is varchar so I cant use - Column(DateTime)
+    publication_date = Column(String(100))
+
+    # (psycopg2.ProgrammingError) column books.lang does not exist
+    # lang = Column(String())
```
Comment (tsauerwein) [inline c2corg_api/models/book.py]: `langs` should be of type `Column(ARRAY(String(2)))`. See also: http://docs.sqlalchemy.org/en/latest/dialects/postgresql.html#sqlalchemy.dialects.postgresql.ARRAY

File: c2corg_api/models/book.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+class _BookMixin(object):
+    author = Column(String(100))
+    editor = Column(String(100))
+    activities = Column(ArrayOfEnum(enums.activity_type))
+    url = Column(String(255))
+    isbn = Column(String(17))
+    book_types = Column(ArrayOfEnum(enums.book_type))
+    nb_pages = Column(SmallInteger)
+
+    # it is varchar so I cant use - Column(DateTime)
```
Comment (tsauerwein) [inline c2corg_api/models/book.py]: This field can be something like ""1982-1998" so it is ok.

File: c2corg_api/scripts/migration/documents/books.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+            activities=self.convert_types(
+                document_in.activities, MigrateRoutes.activities),
+            book_types=self.convert_types(
+                document_in.book_types, MigrateBooks.book_types),
+            author=document_in.author,
+            editor=document_in.editor,
+            url=document_in.url,
+            isbn=document_in.isbn,
+            # langs=document_in.langs,
```
Comment (tsauerwein) [inline c2corg_api/scripts/migration/documents/books.py]: When migrating `langs` we have to decide if we only accept languages defined in [default_langs](https://github.com/c2corg/v6_common/blob/master/c2corg_common/attributes.py#L1) or any language. E.g. "ot".

To get all used languages, you can run:
```
select lang from (select unnest(langs) lang from app_books_archives)t group by lang;
```

File: c2corg_api/views/book.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+class BookRest(DocumentRest):
+
+    @view(validators=[validate_pagination, validate_preferred_lang_param])
+    def collection_get(self):
+        return self._collection_get(BOOK_TYPE, book_documents_config)
+        # BOOK_TYPE - shouldn't be in the request because it causes error
+
+    @view(validators=[validate_id, validate_lang_param])
+    def get(self):
+        return self._get(Book, schema_book)
```
Comment (tsauerwein) [inline c2corg_api/views/book.py]: You can set `include_areas=False` because books should not have a geometry.

File: c2corg_api/search/mappings/book_mapping.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    book_types = QEnumArray(
+        'btyp', model_field=Book.book_types)
+
+    isbn = QEnum(
+        'isbn', model_field=Book.isbn)
+    author = QEnum(
+        'auth', model_field=Book.author)
+    editor = QEnum(
+        'edit', model_field=Book.editor)
+    publication_date = QEnum(
```
Comment (tsauerwein) [inline c2corg_api/search/mappings/book_mapping.py]: Should you be able to search by the publication date?

File: c2corg_api/search/mappings/book_mapping.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    activities = QEnumArray(
+        'act', model_field=Book.activities)
+    book_types = QEnumArray(
+        'btyp', model_field=Book.book_types)
+
+    isbn = QEnum(
+        'isbn', model_field=Book.isbn)
+    author = QEnum(
+        'auth', model_field=Book.author)
+    editor = QEnum(
```
Comment (tsauerwein) [inline c2corg_api/search/mappings/book_mapping.py]: I am not sure about the type for `author` and `editor`. Should these two fields be searchable? In the full-text-search?

File: c2corg_api/views/document_associations.py [hunk 8]
```diff
@@ -12,22 +13,23 @@
...
     route_documents_config, user_profile_documents_config, \
     image_documents_config, article_documents_config, area_documents_config, \
-    outing_documents_config
+    outing_documents_config, book_documents_config
 from c2corg_api.views.validation import updatable_associations
 from sqlalchemy.sql.expression import or_, and_
 
 associations_to_include = {
     WAYPOINT_TYPE: {
         'waypoints', 'waypoint_children', 'routes', 'images', 'articles'},
```
Comment (tsauerwein) [inline c2corg_api/views/document_associations.py]: Waypoints can also be associated to books.

File: c2corg_api/views/document_associations.py [hunk 9]
```diff
@@ -12,22 +13,23 @@
...
     IMAGE_TYPE:
         {'waypoints', 'routes', 'images', 'users', 'outings', 'articles',
-         'areas'},
+         'areas', 'books'},
     ARTICLE_TYPE:
-        {'waypoints', 'routes', 'images', 'users', 'outings', 'articles'},
+        {'waypoints', 'routes', 'images', 'users', 'outings', 'articles',
+         'books'},
     AREA_TYPE: {'images'},
     USERPROFILE_TYPE: {'images'}
```
Comment (tsauerwein) [inline c2corg_api/views/document_associations.py]: If you want that associations are included when requesting the details for a book, you will have to add an entry `BOOK_TYPE: {...}`.

File: c2corg_api/views/document_associations.py [hunk 10]
```diff
@@ -242,5 +247,27 @@ def get_linked_articles(document, lang):
         article_ids, lang, article_documents_config).get('documents')
 
 
+def get_linked_books(document, lang):
+    if document.type in [ARTICLE_TYPE,
```
Comment (tsauerwein) [inline c2corg_api/views/document_associations.py]: Do you need this check `if document.type in [...]`?

File: c2corg_api/models/book.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+class _BookMixin(object):
+    author = Column(String(100))
+    editor = Column(String(100))
+    activities = Column(ArrayOfEnum(enums.activity_type))
+    url = Column(String(255))
+    isbn = Column(String(17))
+    book_types = Column(ArrayOfEnum(enums.book_type))
+    nb_pages = Column(SmallInteger)
+    publication_date = Column(String(100))
```
Comment (asaunier) [inline c2corg_api/models/book.py]: > Do those langs represent that the book is printed (available to buy) in that languages

Yes, "langs" is the language(s) the book is written in.

File: c2corg_api/scripts/migration/documents/books.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    def get_name(self):
+        return 'books'
+
+    def get_model_document(self, locales):
+        return DocumentLocale if locales else Book
+
+    def get_model_archive_document(self, locales):
+        return ArchiveDocumentLocale if locales else ArchiveBook
+
+    def get_document_geometry(self, document_in, version):
```
Comment (tsauerwein) [inline c2corg_api/scripts/migration/documents/books.py]: Is this function used?

File: c2corg_api/models/book.py (hunk 11 above)
Comment (tsauerwein) [inline c2corg_api/models/book.py]: Please remember to add `langs` (also in the migration script).

File: c2corg_api/tests/views/test_book.py [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+                        {'document_id': self.article2.document_id}
+                    ],
+                    'images': [
+                        {'document_id': self.image2.document_id}
+                    ],
+                    'routes': [
+                        {'document_id': self.route2.document_id}
+                    ]
+                }
+                # ,
```
Comment (tsauerwein) [inline c2corg_api/tests/views/test_book.py]: Remove this block if it is not needed. Or did you explicitly want to test that no geometry can be added?

File: c2corg_api/tests/views/test_book.py [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+
+        association_img_log = self.session.query(AssociationLog). \
+            filter(AssociationLog.parent_document_id ==
+                   book1.document_id). \
+            filter(AssociationLog.child_document_id ==
+                   self.image2.document_id). \
+            first()
+        self.assertIsNotNone(association_img_log)
+
+        # check that a link to the associated article is created
```
Comment (tsauerwein) [inline c2corg_api/tests/views/test_book.py]: article?

File: c2corg_api/scripts/migration/documents/books.py (hunk 12 above)
Comment (dkocich) [inline c2corg_api/scripts/migration/documents/books.py]: Yes and no... If I remove it then it crashes during migration -

Traceback (most recent call last):
  File "/home/dkocich/projects/c2corg/v6_api/c2corg_api/scripts/migration/migrate.py", line 93, in <module>
    main()
  File "/home/dkocich/projects/c2corg/v6_api/c2corg_api/scripts/migration/migrate.py", line 83, in main
    MigrateBooks(connection_source, session, batch_size).migrate()
  File "/home/dkocich/projects/c2corg/v6_api/c2corg_api/scripts/migration/documents/document.py", line 76, in migrate
    self._migrate(locales=False)
  File "/home/dkocich/projects/c2corg/v6_api/c2corg_api/scripts/migration/documents/document.py", line 146, in _migrate
    document_in, version)
  File "/home/dkocich/projects/c2corg/v6_api/c2corg_api/scripts/migration/documents/document.py", line 61, in get_document_geometry_archive
    doc['id'] = document_in.document_archive_id
TypeError: 'NoneType' object does not support item assignment

File: c2corg_api/tests/views/test_book.py (hunk 13 above)
Comment (dkocich) [inline c2corg_api/tests/views/test_book.py]: Yes, I want to test geometry like I did also for the articles, which should not have geometry also and we tested it.

File: c2corg_api/scripts/migration/documents/books.py (hunk 12 above)
Comment (tsauerwein) [inline c2corg_api/scripts/migration/documents/books.py]: It looks like this is also a problem with the articles, see https://github.com/c2corg/v6_api/issues/456.

We should make sure that no empty geometries are created for books and articles during the migration.

**Comment by dkocich** [comment]
Note: I should look here before merge ... https://github.com/c2corg/v6_api/blob/de6710114c289072a07bfe02064ee144400e5dda/c2corg_api/tests/views/test_book.py#L377-L390

File: c2corg_api/tests/views/test_book.py [hunk 15]
```diff
@@ -0,0 +1,564 @@
...
+
+        # version with lang 'fr'
+        version_fr = self.get_latest_version('fr', versions)
+        archive_locale = version_fr.document_locales_archive
+        self.assertEqual(archive_locale.title, 'Escalades au Thaurac')
+
+        # check if geometry is not stored in database afterwards
+        self.assertIsNone(book1.geometry)
+
+        # IF I TRY TO USE 'image2' here, then only the LOG part don't work!
```
Comment (tsauerwein) [inline c2corg_api/tests/views/test_book.py]: Did you still plan to add a check for the association with the image?

File: c2corg_api/views/document_associations.py [hunk 16]
```diff
@@ -219,11 +225,12 @@ def get_linked_articles(document, lang):
         Association.parent_document_id == Article.document_id
     )
 
-    if document.type == IMAGE_TYPE:
+    if document.type in [IMAGE_TYPE]:
```
Comment (tsauerwein) [inline c2corg_api/views/document_associations.py]: Why this change?

Comment (dkocich) [inline c2corg_api/views/document_associations.py]: mistake-changed back again

File: c2corg_api/tests/views/test_book.py (hunk 15 above)
Comment (dkocich) [inline c2corg_api/tests/views/test_book.py]: added and deleted comment, it should be ok now.

**Comment by dkocich** [comment]
I added that check for empty geom in articles / books and tested it. I did not have any geometry in new DB and in archives. Then I imported also huts and I had them only for huts, so it should be fine.

File: c2corg_api/views/validation.py [hunk 17]
```diff
@@ -333,7 +331,10 @@ def _add_associations(
 
 
 def _is_parent_of_association(main_document_type, other_document_type):
-    if (main_document_type, other_document_type) in valid_associations:
+    # book is always main_d_t=parent?
+    if main_document_type == 'b' and other_document_type in valid_associations:
```
Comment (tsauerwein) [inline c2corg_api/views/validation.py]: Hmm, I am not sure why you would even need a special case for books. If the associations with books are listed in `valid_associations`, it should be detected properly.

But thinking about it: We should test that associations between books and books are handled correctly.

Comment (dkocich) [inline c2corg_api/views/validation.py]: Ok, my mistake.... I think that now we have no association between Book and Book...?https://github.com/c2corg/camptocamp.org/blob/master/data/sql/tables/db_associations.sql#L20-L71

File: c2corg_api/views/validation.py [hunk 18]
```diff
@@ -317,9 +317,7 @@ def _add_associations(
                 'body', 'associations.' + document_key,
                 'invalid association type')
         else:
-            if main_document_type == 'b':  # book is always main=parent?
-                is_parent = False
-            elif document_key == 'waypoints':
+            if document_key == 'waypoints' and main_document_type != 'b':
```
Comment (tsauerwein) [inline c2corg_api/views/validation.py]: Please use `BOOK_TYPE` instead of `'b'`.

File: c2corg_api/views/validation.py (hunk 17 above)
Comment (tsauerwein) [inline c2corg_api/views/validation.py]: Yes, you are right. I confused books with articles.

File: c2corg_api/views/validation.py [hunk 19]
```diff
@@ -333,6 +331,7 @@ def _add_associations(
 
 
 def _is_parent_of_association(main_document_type, other_document_type):
+    # book is always main_d_t=parent?
```
Comment (dkocich) [inline c2corg_api/views/validation.py]: It does not. I fixed the last two comments.

**Comment by dkocich** [comment]
Close https://github.com/c2corg/v6_api/issues/456

This issue should be ok and closed now for articles/books, but we still need to adjust code if it makes difference in empty geometry of some reports.

**Comment by dkocich** [comment]
so the migration passed without error and geometry table in DB looks ok now

File: c2corg_api/scripts/migration/documents/books.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+            editor=document_in.editor,
+            url=document_in.url,
+            isbn=document_in.isbn,
+            nb_pages=document_in.nb_pages,
+            publication_date=document_in.publication_date,
+            langs=document_in.langs
+        )
+
+    def get_document_locale(self, document_in, version):
+        description, summary = self.extract_summary(document_in.description)
```
Comment (tsauerwein) [inline c2corg_api/scripts/migration/documents/books.py]: With https://github.com/c2corg/v6_api/pull/478 being merged, please do it like for the other documents, like here: https://github.com/c2corg/v6_api/blob/master/c2corg_api/scripts/migration/documents/articles.py#L87-L89

File: c2corg_api/views/book.py [hunk 21]
```diff
@@ -42,14 +43,17 @@ def get(self):
 
     @restricted_json_view(
             schema=schema_create_book,
-            validators=[validate_book_create,
+            validators=[colander_body_validator,
+                        validate_id,
```
Comment (tsauerwein) [inline c2corg_api/views/book.py]: I think this causes the failing unit tests. For `POST` you don't need `validate_id`.

**Comment by tsauerwein** [comment]
Thanks a lot for your continuous work on this, good job! :+1: 

Please merge after Travis has passed.
